### PR TITLE
Some improvements in SortShuffleRead

### PR DESCRIPTION
**Pull request by jerryshao**
1. Use BlockObjectWriter to write spilled data, statistics for memory spilled size and disk spilled size.
2. Manged the request shuffle block and its size locally to fix bugs in incorrectness of size uncompression.
3. Some small changes.

@sryza , would you mind taking a look at this. I've fixed the previous issue and made some other changes. Thanks a lot.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala [hunk 1]
```diff
@@ -146,48 +170,93 @@ private[spark] class SortShuffleReader[K, C](
...
-    val bos = new BufferedOutputStream(fos, fileBufferSize)
 
     // If the remaining unfetched data would fit inside our current allocation, we don't want to
     // waste time spilling blocks beyond the space needed for it.
+    // We use the request size to calculate the remaining spilled size to make sure the
+    // correctness, since the request size is slightly different from result block size because
+    // of size compression.
     var bytesToSpill = unfetchedBytes
-    val blocksToSpill = new ArrayBuffer[MemoryShuffleBlock]()
+    val blocksToSpill = new mutable.ArrayBuffer[MemoryShuffleBlock]()
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala]: Judging from other files, Spark convention appears to be using "ArrayBuffer", not "mutable.ArrayBuffer"

Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala]: Alright, IDEA complains about the previous code, so I changed. I will return to the previous one.

File: core/src/main/scala/org/apache/spark/util/collection/TieredDiskMerger.scala [hunk 2]
```diff
@@ -183,17 +190,38 @@ private[spark] class TieredDiskMerger[K, C](
...
           // Merge the blocks
-          val itrGroup = blocksToRecordIterators(blocksToMerge)
+          val itrGroup = onDiskBlocksToIterators(blocksToMerge)
           val partialMergedItr =
             MergeUtil.mergeSort(itrGroup, keyComparator, dep.keyOrdering, dep.aggregator)
           // Write merged blocks to disk
           val (tmpBlockId, file) = blockManager.diskBlockManager.createTempShuffleBlock()
-          val fos = new BufferedOutputStream(new FileOutputStream(file), fileBufferSize)
-          blockManager.dataSerializeStream(tmpBlockId, fos, partialMergedItr, ser)
+          val curWriteMetrics = new ShuffleWriteMetrics()
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/util/collection/TieredDiskMerger.scala:198]: It looks like we do this in two places.  Can we move the code out into a separate function?

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala [hunk 3]
```diff
@@ -146,48 +170,93 @@ private[spark] class SortShuffleReader[K, C](
...
+          _diskBytesSpilled = file.length()
+        }
       }
-      channel.close()
     }
 
     tieredMerger.registerOnDiskBlock(tmpBlockId, file)
 
-    logInfo(s"Merged ${blocksToSpill.size} in-memory blocks into file ${file.getName}")
+    logInfo(s"Merged ${inMemoryBlocks.size} in-memory blocks into file ${file.getName}")
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala]: Is blocksToSpill not right here?  inMemoryBlocks will hold the blocks that we haven't spilled, right?

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala [hunk 4]
```diff
@@ -198,33 +267,42 @@ private[spark] class SortShuffleReader[K, C](
     }
   }
 
-  private def fetchRawBlocks(): Iterator[(BlockId, Try[ManagedBuffer])] = {
-    statuses = SparkEnv.get.mapOutputTracker.getServerStatuses(handle.shuffleId, startPartition)
+  private def computeShuffleBlocks(): Unit = {
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala:274]: This method should have a header comment explaining what it does.

**Comment by sryza** [comment]
Thanks for the update.  Had a few comments, but otherwise this looks good.

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala (hunk 3 above)
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala]: Oops, I forgot to change this, will modify it.

**Comment by jerryshao** [comment]
Thanks for review, I will address these comments. Besides is there anything we left, I think we can submit a PR to Spark master now.

File: core/src/main/scala/org/apache/spark/util/collection/TieredDiskMerger.scala (hunk 2 above)
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/util/collection/TieredDiskMerger.scala:198]: Yes, this is duplicated in two places, I'm not sure where is suitable to add this function, do you have any suggestion?

**Comment by jerryshao** [comment]
Hi Sandy, I just address all the comments besides the separate function for writing spilled data, I'm not sure where is the best place to put this function, so I didn't change this. Thanks for your reviewing.

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala [hunk 5]
```diff
@@ -146,48 +170,93 @@ private[spark] class SortShuffleReader[K, C](
...
       val block = inMemoryBlocks.dequeue()
       blocksToSpill += block
-      bytesToSpill -= block.blockData.size
+      bytesToSpill -= shuffleBlockMap(block.blockId.asInstanceOf[ShuffleBlockId])._2
     }
 
-    if (blocksToSpill.size > 1) {
+    _memoryBytesSpilled += blocksToSpill.map(_.blockData.size()).sum
+
+    if (inMemoryBlocks.size > 1) {
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala:194]: this should be "if (blocksToSpill.size > 1)", right?

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala [hunk 6]
```diff
@@ -146,48 +170,93 @@ private[spark] class SortShuffleReader[K, C](
...
       }
-      channel.close()
     }
 
     tieredMerger.registerOnDiskBlock(tmpBlockId, file)
 
     logInfo(s"Merged ${blocksToSpill.size} in-memory blocks into file ${file.getName}")
 
-    for (block <- blocksToSpill) {
+    for (block <- inMemoryBlocks) {
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala:252]: This should be blocksToSpill as well, right?

File: core/src/main/scala/org/apache/spark/util/collection/TieredDiskMerger.scala [hunk 7]
```diff
@@ -174,7 +181,7 @@ private[spark] class TieredDiskMerger[K, C](
         }
 
         if (onDiskBlocks.size() > maxMergeFactor) {
-          val blocksToMerge = new ArrayBuffer[DiskShuffleBlock]()
+          val blocksToMerge = new mutable.ArrayBuffer[DiskShuffleBlock]()
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/util/collection/TieredDiskMerger.scala:184]: still prefixes with "mutable"?

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala (hunk 6 above)
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala:252]: That's right, sorry to forget that.

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala (hunk 5 above)
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala:194]: Yes, my fault, I will modify it.

**Comment by jerryshao** [comment]
Hi sandy, I just fix the problems you mentioned before and merged this branch. I also tested with 500G data, seems still has netty managed buffer leakage, do you have any sense where is the possible place?

**Comment by sryza** [comment]
Cool.  What's the evidence that you're seeing of the leakage?

**Comment by jerryshao** [comment]
Here is the log:

```
ERROR ResourceLeakDetector: LEAK: ByteBuf.release() was not called before it's garbage-collected. Enable advanced leak reporting to find out where the leak occurred. To enable advanced leak reporting, specify the JVM option '-Dio.netty.leakDetectionLevel=advanced' or call ResourceLeakDetector.setLevel()
```

Also the real memory occupation is larger than the specified JVM memory, which make system kill the executor when oom.

This is the syslog:

```
java invoked oom-killer: gfp_mask=0x201da, order=0, oom_adj=0, oom_score_adj=0
```

**Comment by sryza** [comment]
Strange.  Nothing sticks out to me for why this could be occurring.  Did you try enabling the advanced leak reporting?  I suppose if there was an exception during the spill we might not release?

**Comment by sryza** [comment]
How large of a job is required to trigger this?

**Comment by jerryshao** [comment]
I haven't yet tried enabling the advanced leak reporting. I will try it later, I guess there might be some corners in the code which makes the release code not executed. I will try to debug it further.

The job size is really depends on the machine's memory size. I have a cluster of 6 nodes with each 48G memory, data size is about 400G with 4000 mappers and 1000 reducers.

**Comment by sryza** [comment]
Is that full data size required to trigger it?  Or can it be triggered in local mode with smaller data?

**Comment by jerryshao** [comment]
Hi Sandy, I have some clues, I'm not sure it is correct or not. Seems Netty memory pool use direct buffer instead of heap buffer, and our SortShuffleRead's buffer holding mechanism is different from HashShuffleRead:
- In HashShuffleRead, Netty ByteBuf is one by one allocated and release after this iterator is out, you can see ShuffleBlockFetcherIterator.
- While in our SortShuffleRead, ByteBuf will be held until shuffle memory is not enough, this requires direct buffer area should be the same as ShuffleMemory's space, and in my node, on-heap memory is allocated for 36G, and the total memory of executor process is about 46G (seeing from Top), So I think additional 10G is for netty ByteBuf.

This may have the problem that shuffle memory manager mechanism actually is not well worked, because it manages heap buffer, which Netty uses direct buffer.

I'm not familiar with Netty, it's just my guess, what's your opinion?

**Comment by jerryshao** [comment]
And I think this will be occurred when memory is critical for this node.

**Comment by sryza** [comment]
Was thinking earlier today that that could be the problem.  I wonder if there's a way to fetch into non-direct buffers?  If not, the easiest fix would be to copy the contents of the direct byte buffers into non-direct byte buffers, right?

**Comment by jerryshao** [comment]
I'm not 100 percent sure about it, this is just my guess. If this is actual the problem, as you said, the easiest way to fix this is to copy the direct buffer into non-direct buffer, but I'm not sure the overhead of this, since the memory bandwidth is quite large, I will modify the code to test again.

**Comment by jerryshao** [comment]
I'm not familiar with Netty, now I'm testing with NIO block transfer service to identify the problem. If NIO is OK, the problem can be focused on Netty buffer allocator.

**Comment by sryza** [comment]
Probably not related, but just in case you didn't see it: https://github.com/apache/spark/pull/3259

**Comment by jerryshao** [comment]
Thanks, I just saw this, I don't think the above PR mentioned issue is the main cause.

I just run another round with NIO, now the memory of executor process keeps with about 38G, not increased like Netty. So I think this direct buffer might be the problem. I will change the code to copy the memory to non-direct buffer and test again.

**Pull request merged**